File: webapi/Models/Flexibility/Efficiency/EfficiencyExercise.cs
using webapi.Models.Math;

namespace webapi.Models.Flexibility
{
    public class EfficiencyExercise : IFlexibilityExercise
    {
        public long? Id { get; set; }

        public int? Ordering { get; set; }

        public FlexibilityExerciseType Type { get; } = FlexibilityExerciseType.Efficiency;

        public bool TransformationRequired { get; set; }

        public bool UseWithTip { get; set; }

        public LinearEquation FirstEquation { get; set; }

        public LinearEquation SecondEquation { get; set; }

        public IsolatedIn FirstEquationIsIsolatedIn { get; set; }

        public IsolatedIn SecondEquationIsIsolatedIn { get; set; }

        public Variable FirstVariable { get; set; }

        public Variable SecondVariable { get; set; }

        public List<Method> EfficientMethods { get; set; }

        public List<SelfExplanation> SelfExplanationTasks { get; set; }

        public string? Question { get; set; }

        public string AgentMessageForSelfExplanation { get; set; }

        public string AgentMessageForFirstSolution { get; set; }

        public string AgentMessageForSecondSolution { get; set; }

        public EfficiencyExercise() { }
    }
}

File: reactapp/src/types/flexibility/efficiencyExercise.ts
import { Type } from "class-transformer";
import { FlexibilityEquation } from "@/types/math/linearEquation.ts";
import { Variable } from "@/types/flexibility/variable.ts";
import { FlexibilityExerciseType, IsolatedIn, Method } from "@/types/flexibility/enums.ts";
import { SelfExplanation } from "@/types/flexibility/selfExplanation.ts";
import { IFlexibilityExercise } from "@/types/flexibility/flexibilityExercise.ts";

export class EfficiencyExercise implements IFlexibilityExercise {
    public readonly id: number;

    public readonly type: FlexibilityExerciseType = FlexibilityExerciseType.Efficiency;

    public readonly transformationRequired: boolean;

    public readonly useWithTip: boolean;

    @Type(() => FlexibilityEquation)
    public readonly firstEquation: FlexibilityEquation;

    @Type(() => FlexibilityEquation)
    public readonly secondEquation: FlexibilityEquation;

    public readonly firstEquationIsIsolatedIn: IsolatedIn;

    public readonly secondEquationIsIsolatedIn: IsolatedIn;

    @Type(() => Variable)
    public readonly firstVariable: Variable;

    @Type(() => Variable)
    public readonly secondVariable: Variable;

    public readonly efficientMethods: Method[];

    @Type(() => SelfExplanation)
    public readonly selfExplanationTasks: SelfExplanation[];

    public readonly question: string | undefined;

    public readonly agentMessageForSelfExplanation?: string | undefined;

    public readonly agentMessageForFirstSolution?: string | undefined;

    public readonly agentMessageForSecondSolution?: string | undefined;

    constructor(id: number, transformationRequired: boolean, useWithTip: boolean, firstEquation: FlexibilityEquation, secondEquation: FlexibilityEquation,
                firstEquationIsIsolatedIn: IsolatedIn, secondEquationIsIsolatedIn: IsolatedIn, firstVariable: Variable,
                secondVariable: Variable, efficientMethods: Method[], selfExplanationTasks: SelfExplanation[], question: string | null,
                agentMessageForSelfExplanation?: string | null, agentMessageForFirstSolution?: string | null, agentMessageForSecondSolution?: string | null) {
        this.id = id;
        this.transformationRequired = transformationRequired;
        this.useWithTip = useWithTip;
        this.firstEquation = firstEquation;
        this.secondEquation = secondEquation;
        this.firstEquationIsIsolatedIn = firstEquationIsIsolatedIn;
        this.secondEquationIsIsolatedIn = secondEquationIsIsolatedIn;
        this.firstVariable = firstVariable;
        this.secondVariable = secondVariable;
        this.efficientMethods = efficientMethods;
        this.selfExplanationTasks = selfExplanationTasks;
        this.question = question === null ? undefined : question;
        this.agentMessageForSelfExplanation = agentMessageForSelfExplanation === null ? undefined : agentMessageForSelfExplanation;
        this.agentMessageForFirstSolution = agentMessageForFirstSolution === null ? undefined : agentMessageForFirstSolution;
        this.agentMessageForSecondSolution = agentMessageForSecondSolution === null ? undefined : agentMessageForSecondSolution;
    }
}